cfep-09.md

```diff
@@ -0,0 +1,100 @@
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+This CFEP proposes a change to how our internal pinning system works.
+The new system will comprise of a new repo for proposed migrations.
+To propose a migration a PR will be issued with a new version of
+the global pinning file and some metadata about the migration needed 
+(if any).
+Once the PR is squashed and merged the bot will take the metadata
+to calculate which packages need to be rebuilt and start issuing
+PRs with the new pinning as a local pinning file to the feedstocks
+in topo order.
+
+This CFEP is taken from @beckermr's work on [conda-forge.github.io](https://github.com/conda-forge/conda-forge.github.io/issues/712).
+
+## Motivation
+Currently when we move pinnings we move the pin in the ``conda-forge-pinning-feedstock``.
+Then we start a migration using the ``regro-cf-autotick-bot``.
+This order of operations causes some issues, when new feedstocks
+are created they recieve the new pinnings even if their dependencies
+are not rebuilt with the new pinnings themselves.
+
+## Specification
+
+### The migrations proceed as follows:
+1. We make a PR on the `migrations` repo with a change to the global pinnings file, generating a new version. 
+2. Once the change is merged, the bot picks it up and starts going through the graph in topo order. 
+3. A migration PR is issued for a node only if 
+    1. The node depends on the changed pinnings.
+    2. It has no dependencies that depend on the new pinnings and have not been migrated.
+    3. It is not currently being migrated by an earlier migration. Per @scopatz, this state is stored in `libcflib` with a REST API.
+4. Process 3 continues until we determine that the migration is sufficiently complete and the change is merged into the global pinnings file. (See below for how a migration is declared complete.)
+
+### Implementation Details:
+- We would need to associate a version independent of git hash with each migration that orders them in time and thus orders the pinning files in time. A pinnings file could be versioned simply with a comment inserted at the top of the YAML containing the timestamp that the change was merged or something. We could use seconds from the epoch!
+- For a given migration, we only use the changes up to and including the commit that has the relevant pinning updates, even if further updates are merged for future migrations. This forces a single global pinning file to be used for a given subgraph that is being migrated. This seems required for consistency.
+- We allow for a local pinnings file to be present in each feedstock that would override the global pinnings if the version is later. In other words, the more recent pinnings file is always preferred by smithy. 
+- Smithy should use the most recent pinnings file when building the `.ci_support` files. This makes sure that rerendering the feedstock always produces the correct pinnings.
```

Comment:
One alternate path here would be to instead store the correct global pinning version for each feedstock in the conda-forge.yaml configuration file. Then upon rerender conda-smithy would need to be changed to read that version and pull the appropriate package. 

This would push all of the work for which pinning file to use to the migration bot.

Reply:
Actually no we should not do this.